### ISSUE TITLE
Support for size in multinomial()

### ISSUE THREAD

**Issue by senarvi** [Theano/Theano#5292]
For implementing sampling based output layers (noise-contrastive estimation, BlackOut, ...), I think it would be important for `multinomial()` to support the `size` argument. The `n` argument cannot be used for sampling multiple "noise" samples, as it will return a distribution of the samples, instead of the individual samples. In order to sample multiple samples, one has to create a 2-dimensional `pvals` matrix, where the same probability distribution is repeated. This is very inefficient if the number of output classes is thousands or millions.

Using `multinomial_wo_replacement()` is a lot more efficient, as it returns a list of the sampled indices. It would be nice if that interface would support sampling _with_ replacement.

**Comment by nouiz** [comment]
@abergeron and @aalmah, what do you think? How frequently this case is?

We could support the broadcast of pvals. We could do this by adding the size parameter. If not set, then we use pvals.size. This would be a new input to the node that is always present. We could allow the broadcast of pvals to size to allow allocating a smaller pvals to support this case.

**Comment by nouiz** [comment]
@lamblin told me that this is what we have in the RandomStream (CPU only) code. So we should support the same interface in both case.

**Comment by lamblin** [comment]
Indeed, see https://github.com/Theano/Theano/blob/140d0a064523349b630a284247c7cddd767fc46e/theano/tensor/raw_random.py#L767 and https://github.com/Theano/Theano/blob/140d0a064523349b630a284247c7cddd767fc46e/theano/tensor/raw_random.py#L840

**Comment by aalmah** [comment]
@nouiz I think this can be done, but I'm not quite sure why we cannot retrieve the actual samples from `n` argument. This argument specified the number samples taken from each `pvals` row (distribution), and we return a (un-normalized) historgram. Can't one just convert this matrix of counts into actual samples? I don't think the order matters...

**Comment by senarvi** [comment]
One could convert the histogram to samples and then randomize the order of the samples. The order doesn't matter as long as it's random.

**Comment by kolasanichaitanya** [comment]
Is someone working on this fix If not I would love  to contribute

**Comment by nouiz** [comment]
To my knowledge, nobody work on this. So you can start it.

thanks!

On Wed, Dec 7, 2016 at 12:11 PM, kolasanichaitanya <notifications@github.com

**Comment by kolasanichaitanya** [comment]
Where is the actual function to change there is a C code label but I
couldn't find related .c source.?

**Comment by nouiz** [comment]
It is in the file theano/sandbox/rng_mrg.py

It is in the methods c_code().

Here is a tutorial on how our c code get generated:

http://deeplearning.net/software/theano/extending/fibby.html

thanks

**Comment by nouiz** [comment]
I discussed this now with @harmdevries89. The title of this issue and the description ask for 2 different things:
- (the title) Add support for size parameter in multinomial() method to allow broadcasting of the pvals as ours numpy multinomial wrapper. multinomial return the histogram of the sample.
- (the description) Add support in multinomial_wo_replacement() to do with replacement. multinomial_wo_replacement() return the sample indices.

From the description, for your use cases, the second would be the most efficient. I think it would be small code changes in the c code and perform of those classes (CPU and GPU). So it make sense to do that. @harmdevries89 raised that we will need to change the class name to prevent confusion.

@aalmah @lamblin @abergeron why multinomial_wo_replacements was made to don't return the output as multinomial? Do we need both output format or the multinomial output was always converted to the indices after?

@senarvi do you agree with what I wrote?

**Comment by senarvi** [comment]
I don't know how the functions are implemented, but I would guess that modifying `multinomial_wo_replacement()` is easier and indeed more efficient, as I need the sampled values, not a histogram.

**Comment by nouiz** [comment]
While talking with @lamblin he raised that it is not normal that multinomial_wo_replacement return the index. This was an interface mistake. It should have been called choice() as in numpy. I made an issue about this https://github.com/Theano/Theano/issues/5408 and that include the optimization.

So what is left for this issue is to make size work with multinomial in MRG and make broadcast work like in our numpy wrapper.
